rpc/query_server_gateway.go

```diff
@@ -0,0 +1,132 @@
+// +build gateway
+
+package rpc
+
+import (
+	"github.com/loomnetwork/go-loom"
+	"github.com/loomnetwork/go-loom/plugin/contractpb"
+	"github.com/loomnetwork/loomchain/builtin/plugins/ethcoin"
+	"github.com/pkg/errors"
+
+	amtypes "github.com/loomnetwork/go-loom/builtin/types/address_mapper"
+	tgtypes "github.com/loomnetwork/go-loom/builtin/types/transfer_gateway"
+	glcommon "github.com/loomnetwork/go-loom/common"
+	gtypes "github.com/loomnetwork/go-loom/types"
+	am "github.com/loomnetwork/loomchain/builtin/plugins/address_mapper"
+	loomcoin "github.com/loomnetwork/loomchain/builtin/plugins/coin"
+	"github.com/loomnetwork/transfer-gateway/builtin/plugins/gateway"
+)
+
+func (s *QueryServer) GetAccountBalances(contracts []string) (*AccountsBalanceResponse, error) {
+	snapshot := s.StateProvider.ReadOnlyState()
+	defer snapshot.Release()
+
+	addrMapperCtx, err := s.createAddressMapperCtx(s.StateProvider.ReadOnlyState())
+	if err != nil {
+		return nil, err
+	}
+	var resp *amtypes.AddressMapperListMappingResponse
+	resp, err = am.ListMapping(addrMapperCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Mappings) == 0 {
+		return nil, errors.Errorf("Empty address mapping list")
+	}
+
+	ethcoinCtx, err := s.createStaticContractCtx(snapshot, "ethcoin")
+	if err != nil {
+		return nil, err
+	}
+
+	loomCoinCtx, err := s.createStaticContractCtx(snapshot, "coin")
+	if err != nil {
+		return nil, err
+	}
+
+	gatewayCtx, err := s.createStaticContractCtx(snapshot, "gateway")
+	if err != nil {
+		return nil, err
+	}
+
+	accounts := make(map[string]map[string]string, len(resp.Mappings))
+	for _, mp := range resp.Mappings {
```

Comment:
Since mappings are bi-directional there'll be 2 entries for each account, so you'll be fetching all the balances twice. Should check if `accounts` already has an entry for this account before fetching balances.